### PR TITLE
[ci skip] Add console examples for each association type.

### PR DESCRIPTION
**Pull request by marklocklear**
For kinesthetic learners (doers) it is important to show concrete examples of how advanced concepts like associations work. What I have attempted to do here is (using the console) give examples of how each of the associations works based on the models and attributes for each. Existing text/examples were left in place and supplemented with console examples. Note that I did remove the published_at attribute in 2.1 belongs_to and 2.3 has_many and replace with :name as it seems to make more sense for the examples I am using. I also updated the diagrams/images that went along with these two examples.

### CODE REVIEW

File: guides/source/association_basics.md [hunk 1]
```diff
@@ -110,13 +110,37 @@ class CreateBooks < ActiveRecord::Migration[5.0]
 
     create_table :books do |t|
       t.belongs_to :author, index: true
-      t.datetime :published_at
+      t.string :name
```
Comment (maclover7) [inline guides/source/association_basics.md]: Looks like when the graphs were regenerated, the arrows are now pointing from the wrong box 😬

File: guides/source/association_basics.md [hunk 2]
```diff
@@ -110,13 +110,37 @@ class CreateBooks < ActiveRecord::Migration[5.0]
...
+```
+
+Now we can associate this author with a book:
+
+```ruby
+@book = Book.create(name: "Remote: Office Not Required", author_id: @author.id)
+ => #<Book id: 1, author_id: 1, name: "Remote: Office Not Required", ...>
+```
+
+Note that we inserted the id of the author David Heinemeier Hansson into the `author_id` field of our book. With this association in place we can view this assocation from the perspective of the book:
```
Comment (maclover7) [inline guides/source/association_basics.md]: Can you add a comma after association

Comment (maclover7) [inline guides/source/association_basics.md]: Can you put the author's name in parentheses

File: guides/source/association_basics.md [hunk 3]
```diff
@@ -110,13 +110,37 @@ class CreateBooks < ActiveRecord::Migration[5.0]
...
       t.belongs_to :author, index: true
-      t.datetime :published_at
+      t.string :name
       t.timestamps
     end
   end
 end
 ```
 
+As an example, using the rails console, let's first create an author:
```
Comment (maclover7) [inline guides/source/association_basics.md]: Can you try and standardize the top language; here it's "using the rails console", below it's "working from the console", etc...

File: guides/source/association_basics.md [hunk 4]
```diff
@@ -158,10 +182,34 @@ create_table :accounts do |t|
...
+ => #<Supplier id: 1, name: "Acme", ...>
+```
+
+Now we can create an account that is automatically associated with this supplier:
+
+```ruby
+@account = @supplier.create_account(account_number: '123')
+ => #<Account id: 1, supplier_id: 1, account_number: "123", ...>
+```
+Note that the `supplier_id` (in this case 1) is inserted into the Accounts table. Now let's add a second account_number for this supplier:
```
Comment (maclover7) [inline guides/source/association_basics.md]: Can you downcase accounts, and wrap it with tick marks

Comment (maclover7) [inline guides/source/association_basics.md]: Can you add tick marks around account_number

File: guides/source/association_basics.md [hunk 5]
```diff
@@ -158,10 +182,34 @@ create_table :accounts do |t|
...
+
+```ruby
+@account = @supplier.create_account(account_number: '456')
+  INSERT INTO "accounts" [["supplier_id", 1], ["account_number", "456"], ...]
+  UPDATE "accounts" [["supplier_id", nil], ..., ["id", 1]]
+ => #<Account id: 2, supplier_id: 1, account_number: "456", ...>
+```
+Now we see the true power of the `has_one` association. Notice that when we inserted another account_number for this same supplier into the database that two things happened. First, we inserted a new record in to the accounts table with an account_number of 456. Second, this command also updated the previous record we inserted into the database and set the supplier_id to nil. This database transaction ensures that a supplier truly only "has one" account.
```
Comment (maclover7) [inline guides/source/association_basics.md]: in to --> into

File: guides/source/association_basics.md [hunk 6]
```diff
@@ -252,6 +316,32 @@ If some that existed previously are now missing, then their join rows are automa
 
 WARNING: Automatic deletion of join models is direct, no destroy callbacks are triggered.
 
+To see this association in action let's create a physician and a couple of patients:
```
Comment (maclover7) [inline guides/source/association_basics.md]: Can you add a comma after action

File: guides/source/association_basics.md [hunk 7]
```diff
@@ -252,6 +316,32 @@ If some that existed previously are now missing, then their join rows are automa
...
+
+Now let's create appointments to associate the patients to our physician:
+
+```ruby
+@patient1.appointments.create(physician: @physician, appointment_date: Date.today)
+ => #<Appointment id: 1, physician_id: 1, patient_id: 1, appointment_date: "2017-03-17 00:00:00", ...>
+@patient2.appointments.create(physician: @physician, appointment_date: Date.today)
+ => #<Appointment id: 2, physician_id: 1, patient_id: 2, appointment_date: "2017-03-17 00:00:00", ...>
+```
+Note that an appointment contains both a physician and a patient, so Appointments acts as a join table. To see all patients assigned to a physician we can do:
```
Comment (maclover7) [inline guides/source/association_basics.md]: Can you wrap appointments in tick marks

File: guides/source/association_basics.md [hunk 8]
```diff
@@ -325,6 +415,42 @@ class CreateAccountHistories < ActiveRecord::Migration[5.0]
...
+
+ Let's create another account history record:
+
+```ruby
+@supplier.account.create_account_history(credit_rating: 789)
+  INSERT INTO "account_histories" [["account_id", 1], ["credit_rating", 789], ...]
+  UPDATE [["account_id", nil], ["updated_at", 2017-03-22 14:54:40 UTC], ["id", 1]]
+ => #<AccountHistory id: 2, account_id: 1, credit_rating: 789, created_at: "2017-03-22 14:54:40", updated_at: "2017-03-22 14:54:40">
+```
+As in our earlier [`has_one`](##the-has-one-association) example there is an insert statement for the new record that has a credit_rating of 789, but also notice the update statement that sets account_id in the previous record we inserted to nil. We can confirm this with the following console command:
```
Comment (maclover7) [inline guides/source/association_basics.md]: Comma after example

**Review by maclover7 — CHANGES_REQUESTED** [review]
Did a first pass over the changes -- overall looks good. Left some grammar and stylistic comments.

File: guides/source/association_basics.md (hunk 1 above)
Comment (marklocklear) [inline guides/source/association_basics.md]: I am not sure I understand this one. The other suggestions look good. I will work on this and resubmit.

File: guides/source/association_basics.md (hunk 2 above)
Comment (marklocklear) [inline guides/source/association_basics.md]: A comma seems more appropriate here after place; as in " With this association in place, we can view this ". No?

**Review by marklocklear — COMMENTED** [review]
@maclover7 have made changes based on your review.  See also a couple of individual comments below.

**Comment by mjhoy** [comment]
@marklocklear FYI: I noticed that some graphs were changed and have arrows now pointing from the wrong column (see the github diff).

**Comment by marklocklear** [comment]
@mjhoy great catch! Just fixed and pushed. Tried to squash commits but ran into issues. If this is a problem someone shout and I will take another stab at it.

**Comment by mjhoy** [comment]
@marklocklear has_many is fixed but it looks like belongs_to.png is still incorrect.

**Comment by marklocklear** [comment]
@mjhoy updated belongs_to image. Me thinks my commit history is not right. Let me know if I need to clean up. Thanks!

**Comment by mjhoy** [comment]
@marklocklear Yep, something is really off here!

I think you ended up with two branches from this that are both merged in somehow. Perhaps you did a `git pull origin master` against your branch once or twice?

I believe you just need to reset your branch head to `d2ce1d8`, and then force update the remote branch. You might also need to rebase against rails master. Let me know if you need help with this.

If your git index and working tree are "clean" i.e. everything is committed or stashed, I would run:

```sh
# Assuming you are on your association_doc_updates branch
git reset d2ce1d8 # set HEAD to before all the merges
git log # Inspect the history and make sure this is correct
# Maybe: git rebase [RAILS_REMOTE]/master
git push [REMOTE] association_doc_updates -f # force update the remote branch
```

**Comment by mjhoy** [comment]
@marklocklear Also, if you haven't seen it, it's worth checking out the Rails Guides on pull requests and git usage, which has a good, simple workflow you can use. http://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#update-your-branch

**Comment by marklocklear** [comment]
@mjhoy thanks for being patient. I think commit history is clean now.

File: guides/source/association_basics.md (hunk 1 above)
Comment (marklocklear) [inline guides/source/association_basics.md]: This is fixed now. See updated images.

**Comment by marklocklear** [comment]
Hey folks, please let me know if there is anything else I need to do regarding this pull request. Thanks!

**Pull request closed without merging**